### PR TITLE
Enable Minimal ARM64 CI Tests

### PR DESCRIPTION
**Pull request by HernanGatta**
## Overview

During CI runs, the SDK is built for OP-TEE on ARM TrustZone, including a number of tests in its test suite. However, no tests are executed even though some are known to pass on this platform.

This PR enables 17 passing tests to execute during CI runs for ARM64.

### Test Listing

```
Test project /openenclave/build
      Start  1: tests/mem
 1/17 Test  #1: tests/mem ........................   Passed    0.04 sec
      Start  2: tests/safecrt
 2/17 Test  #2: tests/safecrt ....................   Passed    0.21 sec
      Start  3: tests/safemath
 3/17 Test  #3: tests/safemath ...................   Passed    0.09 sec
      Start  4: tests/str
 4/17 Test  #4: tests/str ........................   Passed    0.04 sec
      Start  5: tests/logging
 5/17 Test  #5: tests/logging ....................   Passed    0.03 sec
      Start  6: tests/create-errors
 6/17 Test  #6: tests/create-errors ..............   Passed    0.04 sec
      Start  7: tests/crypto/enclave
 7/17 Test  #7: tests/crypto/enclave .............   Passed   14.08 sec
      Start  8: tests/crypto/host
 8/17 Test  #8: tests/crypto/host ................   Passed    0.47 sec
      Start  9: tests/hexdump
 9/17 Test  #9: tests/hexdump ....................   Passed    0.47 sec
      Start 10: tests/initializers
10/17 Test #10: tests/initializers ...............   Passed    0.20 sec
      Start 11: tests/mixed_c_cpp
11/17 Test #11: tests/mixed_c_cpp ................   Passed    0.17 sec
      Start 12: tests/pingpong
12/17 Test #12: tests/pingpong ...................   Passed    0.17 sec
      Start 13: tests/pingpong-shared
13/17 Test #13: tests/pingpong-shared ............   Passed    0.19 sec
      Start 14: tests/SampleApp
14/17 Test #14: tests/SampleApp ..................   Passed    0.24 sec
      Start 15: tests/SampleAppCRT
15/17 Test #15: tests/SampleAppCRT ...............   Passed    0.19 sec
      Start 16: tests/stdc
16/17 Test #16: tests/stdc .......................   Passed    3.29 sec
      Start 17: tests/libc
17/17 Test #17: tests/libc .......................   Passed    7.07 sec

100% tests passed, 0 tests failed out of 17

Total Test time (real) =  28.11 sec
```

### Dependencies

This PR depends on PR #2526.

## Implementation

To test the SDK's implementation on OP-TEE on ARM TrustZone, an ARM TrustZone-capable device with the appropriate OP-TEE firmware is necessary. Fortunately, [QEMU](https://www.qemu.org/) supports ARM TrustZone emulation.

As part of PR #2526, a build of QEMU with ARM TrustZone support, OP-TEE firmware, the Linux kernel, and a Linux user-mode root filesystem are included in Open Enclave's Full Docker container images. This root filesystem is configured to automatically generate SSH keys during boot and to expose an SSH server.

After the SDK build step, QEMU is started in the background. Via command line parameters, QEMU is configured to set up [user networking](https://wiki.qemu.org/Documentation/Networking#User_Networking_.28SLIRP.29) between the emulated system (i.e., the guest) and its parent Docker container (i.e., the host), with [SSH port forwarding](https://wiki.qemu.org/Documentation/Networking#How_to_get_SSH_access_to_a_guest). Additionally, the host's CMake workspace (that is, the SDK's source code and its build output) are exposed to the guest via [Plan 9 file sharing](https://wiki.qemu.org/Documentation/9psetup). Lastly, two [virtual serial ports](https://qemu.weilnetz.de/doc/qemu-doc.html#Debug_002fExpert-options) are attached to the guest, one for Linux, and one for OP-TEE. The output written to the serial ports is stored in the host's filesystem as `ree.log`, and `tee.log`, respectively.

After QEMU is started, it is necessary to wait briefly until the port forwarding mechanism becomes active (`nc`). Once up, a command (`ssh-keyscan`) is started to extract and accept the auto-generated SSH keys of the guest. This command waits several seconds while the guest boots.

After the keys are extracted and marked as trusted, a command sequence is constructed. This sequence is forwarded to the guest via an SSH connection (`sshpass`).

The command sequence passed to the guest first mounts the shared portions of the host's filesystem, mirroring the host's filesystem layout of the SDK's source and build directories in the guest. This satisfies `ctest`'s usage of absolute paths when running the test suite.

Once this step is complete, the enclaves (`.ta` files) are copied from the build tree into the appropriate location inside the guest so that the TEE supplicant may find them. Thereafter, `ctest` is run in the usual way.

The exit code of `ctest` is forwarded to the host by `sshpass` and is captured in the test script. The Linux and OP-TEE logs are dumped, and `ctest`'s exit code is returned as the exit value for the Jenkins task, indicating success or failure.

**Note:** Once the script finishes, the Docker container is automatically cleaned up, terminating QEMU along with it. Hence, it is not necessary to stop QEMU in the script.

### SDK Changes

To enable `ctest` to run tests on OP-TEE, it is necessary for the `add_test` macro to pass to test hosts the UUID of their enclave instead of the path to it. A change was thus made to `add_enclave` in the OP-TEE case to attach an `ENCLAVE_UUID` property to the enclave target. The `add_test` function reads this property from the enclave target and has it passed to the test host when the SDK is tested on OP-TEE.

Additionally, some tests were modified to support OP-TEE by including the appropriate usage of `OE_SET_ENCLAVE_OPTEE`, and using `OE_ENCLAVE_TYPE_AUTO` in the host instead of the restrictive `OE_ENCLAVE_TYPE_SGX` when creating the enclave.

### Performance

The testing mechanism proposed in this PR takes ~40s +/- 2s on a ThinkPad T490:

* CPU: [Intel Core i7 8665U](https://ark.intel.com/content/www/us/en/ark/products/193563/intel-core-i7-8665u-processor-8m-cache-up-to-4-80-ghz.html)
   1. 4c / 8t
   1. Base CLK = 1.90GHz
   1. Turbo CLK = 4.80GHz (single core)
   1. PL2 = 29W, PL1 = 25W
* Graphics: iGPU only
* Memory: 32GB DDR4 (2x16GB dual channel)
* Storage: Samsung 970 EVO 500GB NVMe
* OS: Xubuntu 19.10 (Linux 5.3.0-29-generic)

**Note:** The emulator is configured to only emulate one core (`-smp 1`), and tests run sequentially; thus, single-threaded performance dominates. Further testing would be required to enable multi-threaded emulation and parallel testing, and to ensure that the system is stable.

## Merging Strategy

To ascertain that this testing system functions before merging it, I propose the following:

1. Rebuild the Docker images:
   1. Use the branch underlying #2526;
   1. Do not tag with `latest`.
1. Execute three SDK CI runs:
   1. Target this PR's base branch;
   1. Explicitly use the custom Docker images;
   1. Use multiple runs to attempt to catch failures due to ephemeral conditions.
1. If success:
   1. Merge #2526;
   1. Rebuild the Docker images:
      1. Tag with `latest`.
   1. Merge this PR.

**Note:** This CI testing strategy was used successfully while the `feature.new_platforms` branch was under development, running via an Azure DevOps Pipeline.

## Further Work

There are other tests in the test suite that might pass as well on OP-TEE on ARM TrustZone. These must be looked into individually. Subsequent PRs will be issued to enable them, too.

Additionally, some tests could pass on this platform but will not due to the assumption some tests make that the enclave can read host memory (e.g., `hostcalls`).

Moreover, building the `samples` directory as part of the test suite is disabled for OP-TEE on ARM TrustZone since the emulated environment does not yet contain the necessary prerequisites to build them.

Lastly, it would be useful going forward to:
1. Not write tests that presume enclaves can read host memory;
2. Utilize `OE_ENCLAVE_TYPE_AUTO` instead of `OE_ENCLAVE_TYPE_SGX`, unless the test is SGX-specific;
3. Call `OE_SET_ENCLAVE_OPTEE`, and set the heap and stack size the same as for SGX;
4. Attempt to run the test on OP-TEE on ARM TrustZone by default, unless the feature under test is known to not be implemented on that platform.

**Note:** I'll produce documentation describing how to retrieve and use the emulated environment manually to run tests locally.

### Open Questions

1. Would you like that the script and patches used to build the emulated environment be committed?
   1. If so, into the SDK tree, into the [CI repository](https://github.com/openenclave/openenclave-ci), or somewhere else?

### CODE REVIEW

**Comment by radhikaj** [comment]
@HernanGatta Thanks for this PR. I think the scripts and patches used to build the emulated environment can be submitted to the SDK tree if they depend on code in the SDK tree. If they are independent, please put them in a repo under github.com/openenclave.

**Comment by jhand2** [comment]
Taking a look at the test run/output

bors try

**Comment by BRMcLaren** [comment]
bors try

**Comment by BRMcLaren** [comment]
Hello @HernanGatta 

Thanks for your contribution!

Commenting to say that the merge strategy outlined is the one we currently use, so yes. 👍 

You fit into the use case of a dependency change instead of an SDK code change, so we need to run through this process when you are ready, as such I am adding the "do-not-merge" until me, @ionutbalutoiu  or @oprinmarius can run it through the process.

We are adding a pipeline to do this in the future, just not ready yet.

In the meantime, @HernanGatta Could you do me a favour and merge your PR's into a separate branch for me when you are ready? Don't make a PR, we just need a point of reference to rebuilding images, VHD's and agents from to do an E2E test. Could you also change the tag the images are pulling in something other than latest? maybe ARM64  and when we build the images we will build to match.

Here is an example when we switched our agent configuration:
https://github.com/openenclave/openenclave/commit/87c547c4a87bd1c1506121503ddd0755860355ca

File: .jenkins/Jenkinsfile
Comment (jhand2) [inline .jenkins/Jenkinsfile]: does this one need to be escaped too?

Comment (HernanGatta) [reply .jenkins/Jenkinsfile]: You are correct.

**Comment by HernanGatta** [comment]
> In the meantime, @HernanGatta Could you do me a favour and merge your PR's into a separate branch for me when you are ready? Don't make a PR, we just need a point of reference to rebuilding images, VHD's and agents from to do an E2E test. Could you also change the tag the images are pulling in something other than latest? maybe ARM64 and when we build the images we will build to match.

As requested, here is a branch with the contents of both #2526 and #2527, as well as customer Docker image tags:

https://github.com/HernanGatta/openenclave/tree/user/hegatta/arm-ci-full

**Comment by HernanGatta** [comment]
bors try

**Comment by HernanGatta** [comment]
> @HernanGatta Thanks for this PR. I think the scripts and patches used to build the emulated environment can be submitted to the SDK tree if they depend on code in the SDK tree. If they are independent, please put them in a repo under github.com/openenclave.

@radhikaj: The scripts to create the emulator tarball do not depend on code in the SDK. I don't have enough admin power to create a new repo under `github.com/openenclave`, though. Could you do this on my behalf? Or would you prefer I reuse the `openenclave-ci` repo? Thanks!

**Comment by BRMcLaren** [comment]
bors try

**Comment by radhikaj** [comment]
> openenclave-ci

Yes, please put them in openenclave-ci if the scripts and patches are meant to enable a ci environment.

**Comment by HernanGatta** [comment]
The failure is as expected, since the Docker images don't yet have the emulator in them.

**Comment by HernanGatta** [comment]
> Yes, please put them in openenclave-ci if the scripts and patches are meant to enable a ci environment.

@radhikaj , thank you, [opened a PR there](https://github.com/openenclave/openenclave-ci/pull/10).

**Comment by radhikaj** [comment]
@BRMcLaren  Are we waiting on adding this to the CI/CD pipeline? Can we do this in the 0.9 timeframe?

**Comment by BRMcLaren** [comment]
@radhikaj That is correct, there were unforeseen delays. 0.9 should be fine from a CI/CD perspective.

https://github.com/openenclave/openenclave/pull/2687/files Enables the testing of this.

**Comment by HernanGatta** [comment]
@BRMcLaren:

> As requested, here is a branch with the contents of both #2526 and #2527, as well as custom Docker image tags:
> 
> https://github.com/HernanGatta/openenclave/tree/user/hegatta/arm-ci-full

I've updated that branch with the up-to-date merge of `user/hegatta/arm-ci` and `user/hegatta/arm-ci-prereqs` after I resolved the merge conflicts that appeared in the intermi.

I noticed that you added the `{DOCKER_TAG}` parameter, so that branch no longer has a custom image tag.

**Comment by kuanggang-git** [comment]
Can one of the admins verify this patch?

**Comment by dthaler** [comment]
bors try

**Review by achamayou — APPROVED** [review]

**Comment by radhikaj** [comment]
@HernanGatta  - Please reopen when you can work on this again

**Pull request closed without merging**
